cunumeric/module.py

```diff
@@ -5327,6 +5336,630 @@ def nancumsum(
     )
 
 
+def _get_non_nan_reduction_code_if_applicable(
+    dtype: npt.DTypeLike, unary_reduction_code: UnaryRedCode
+) -> UnaryRedCode:
+    """
+    Return the equivalent non-nan reduction op codes if the datatype of the
+    array doesn't belong to one of the allowed datatypes. Raise an error if the
+    datatype is disallowed, which is currently complex64 and complex128.
+    """
+
+    equivalent_non_nan_unary_red_ops: Dict[UnaryRedCode, UnaryRedCode] = {
+        UnaryRedCode.NANARGMAX: UnaryRedCode.ARGMAX,
+        UnaryRedCode.NANARGMIN: UnaryRedCode.ARGMIN,
+        UnaryRedCode.NANMAX: UnaryRedCode.MAX,
+        UnaryRedCode.NANMIN: UnaryRedCode.MIN,
+    }
+
+    assert unary_reduction_code in equivalent_non_nan_unary_red_ops
+
+    # raise error for disallowed datatypes
+    disallowed_dtypes: list[np.dtype[Any]] = [
+        np.dtype(np.complex64),
+        np.dtype(np.complex128),
+    ]
+    if dtype in disallowed_dtypes:
+        raise NotImplementedError(
+            "operation is not supported for complex64 and complex128 types"
+        )
+
+    # use non-NaN API if the datatype is not floating point type
+    allowed_dtypes: list[np.dtype[Any]] = [
+        np.dtype(np.float16),
+        np.dtype(np.float32),
+        np.dtype(np.float64),
+    ]
+    if dtype in allowed_dtypes:
+        reduction_code = unary_reduction_code
+    else:
+        reduction_code = equivalent_non_nan_unary_red_ops[unary_reduction_code]
+
+    return reduction_code
+
+
+def _handle_exceptions_and_update_out_if_needed(
+    out: ndarray,
+    op: UnaryRedCode,
+    identity: Union[int, np.floating[Any]],
+) -> None:
+    """Raise ValueError if NaN is found in a slice or in the entire
+    array for nanargmin and nanargmax. In case of nanmin and nanmax,
+    issue a RuntimeWarning.
+    """
+    supported_ops = [
+        UnaryRedCode.NANMIN,
+        UnaryRedCode.NANMAX,
+        UnaryRedCode.NANARGMIN,
+        UnaryRedCode.NANARGMAX,
+    ]
+
+    if op not in supported_ops:
+        return
+
+    if identity in out:
+        if op in [UnaryRedCode.NANMIN, UnaryRedCode.NANMAX]:
+            runtime.warn(
+                "Array/Slice contains only NaNs", category=RuntimeWarning
+            )
+
+            # replace identities with NaNs since numpy returns NaNs
+            if out.ndim == 0:
+                out.fill(np.nan)
+            else:
+                mask = out == identity
+                putmask(out, mask, np.nan)  # type: ignore
+        else:
+            raise ValueError("Array/Slice contains only NaNs")
```

Comment:
though this is necessary to match numpy's behavior, but I'd actually remove this and leave a comment about the deviation. a consequence of code like this is that this would slow down the code even though the majority of cases won't actually hit this; the inclusion check at 5400 would block on the computation of `out` and another task for the inclusion check itself. the reason that numpy can accurately report the all-nan case is only because it runs sequentially, which I'd consider an implementation artifact. so we should be a bit more careful in deciding how much of the numpy behavior we want cunumeric to match.

Reply:
also note that your logic is faulty. the array might only have identities, in which case your logic would think the array was filled with nans. in general you can't really check this property accurately without augmenting the computation itself with the check.

Reply:
The initial implementation did not have the part where we replace nans with identities. I figured it'd be okay to return identity instead of NaNs for perf reasons but Irina felt quite strongly about result differences (numpy returns nans we return identity), so I had a fill/putmask to update identities with NaNs to comply with numpy's behavior. I've reverted that change in the latest commit, so we now return identity if we find the array is full of NaNs or when a slice is full of NaNs.

But in general, I think the inclusion check in the beginning of the function to check for identity might be necessary - otherwise users wouldn't know when they're encountering a case when the array is full of NaNs. I'd guess that most users use the nan equivalent of the unary reduction ops because they except that the arrays could have NaNs, so giving them a heads up that something's not kosher might actually be better especially when we are not complying with numpy (numpy will return NaN, we return identity; we can definitely note that deviation in the doc). Also, if we don't raise a warning, the user might just as well use the non-nan API which would return a NaN without a warning.

e.g.:

```
x = np.array([np.nan, np.nan]) 
np.nanmax(x) # returns identity but doesn't warn that array is full of nans
np.max(x)    # returns nan and doesn't warn that array is full of nans (expected behavior)
```

Regarding the case you mentioned when the array is filled with just identity - I've mentioned this in the description that this is not addressed in this PR (check bullet 2 under "corner cases") and have an example where the existing non-nan API might return something incorrect (can be considered a bug but we know it's the expected behavior based on our implementation). The implementation for nan-APIs are based off of the equivalent non-nan APIs, so I figured it'd be okay to leave out that corner case, and not address that here. Other than that case, this inclusion check logic should work. Please let me know if you find any other case where it doesn't work.

Reply:
The most troubling thing about the inclusion check as you've written currently is that it is used all the time, even though it would only ever be needed for rare corner cases. This might have been okay if cunumeric was yet another sequential NumPy implementation, but it isn't and we should be very careful whenever we introduce a blocking behavior like this in the library implementation. So at the very minimum we should make the check optional so it won't impose constant overhead on common scenarios. Also, if I understand your code correctly, the inclusion check is performed even when the output has an integer type. (you might argue who would call a nan-variant on integer inputs, but that can actually happen in a code written in a type generic way.)

And speaking of caching all-NaN cases, I think it's inaccurate and backward to conclude that the input had all NaNs by seeing if the output has any identity. If you really wanted to catch that case, you would have added an extra boolean output from the reduction task indicating that the inputs have all NaNs. And I don't think we would want to do the nan recovery as written currently on axis-wise reduction results, as it would try to go over the entire output even when only one of the rows encountered the all-NaN case.

to sum up, I feel catching and recovering from the all-NaN case isn't the right kind of complication we want to introduce.

Reply:
so it looks like there are three things - blocking behavior, current implementation less optimal for axis-wise reductions, and that the current checking of all-NaNs is backward and inaccurate;

- blocking behavior: I'm assuming here that you are saying that the runtime will be blocked on the result for CONTAINS but not for other reduction operations, did I understand that right? I ask this because, if we were to make an  alternative implementation to figure out if we have all NaNs, that wouldn't be quite straightforward - naively, we could count the elements with NaNs along an axis and do a reduction on the count, and compare the output with the total number of elements along that axis. That would tell us if we had all NaNs along an axis, but would still incur a reduction, and now going back to the question on blocking, since we do this reduction in a non-blocking way, this is better. Is that right?

 - axis-wise reductions not optimal:  Agreed. This is suggesting a much more involved cpp implementation which has been avoided in the current implementation (no optimizations were planned for this PR as well; that was the scope of this PR). 
 
 - current checking of all-NaNs is backward and inaccurate: We could definitely try to implement the check for all NaNs in the computation itself, but I think the backwardness that you are referring is because we know we update the output to identity, so it's really relying on that info, and that's why it fails when the array contains only identities (which fails even for some non-nan APIs, just fyi). As far as the inaccuracy is concerned, I think that's the only case where I can imagine this failing.
 
> to sum up, I feel catching and recovering from the all-NaN case isn't the right kind of complication we want to introduce.

In principle, for most use cases, that is the main benefit of using nan unary reduction operations as opposed to their non-nan equivalents in cuNumeric. Consider an example like this in `cunumeric`:

```

>>> x
array([nan, 0.1])
>>> np.min(x)
array(0.1)
>>> np.nanmin(x)
array(0.1)
>>> np.argmin(x)
array(1)
>>> np.nanargmin(x)
array(1)

```
The results are correct even without the sanitization in some of the non-nan APIs because I think they get sanitized as a consequence of what we do in fold/apply because any comparison check with NaN would fail. If we don't handle the exceptions for say, nanmin and nanmax, we might as well just call cunumeric's min and max for those two APIs (and similarly for argmin and argmax).

Because of that, I'd suggest that we go with issuing warnings as currently implemented, but perhaps restrict under a flag like you were suggesting, and reflecting this in the doc. Does that sound good to you? Or do you still feel strongly about not handling exceptions the way it is implemented in this PR.

Reply:
> Because of that, I'd suggest that we go with issuing warnings as currently implemented, but perhaps restrict under a flag like you were suggesting, and reflecting this in the doc. Does that sound good to you?

No, it doesn't. What I was alluding to was we should just get rid of the inclusion check altogether, as it's inaccurate and causing penalty due to its blocking behavior. If we do want to have that check, then the check must be accurate. What that check is doing is basically converting an algorithm that produces an "incorrect" result (quotes here as it's not really incorrect) for the all-NaN case to another that produces an incorrect result for the all-identity case. Why do you think this is an improvement or acceptable? Again, if we do want to make the algorithm behave in the same way as NumPy's, we should pay attentions to the details so we don't introduce another corner case like that. Whether we should do it or not is a separate question, which I'm inclined to saying no. That said, I wouldn't stop you from writing a correct one. But the current inclusion check shouldn't be included.

> blocking behavior: I'm assuming here that you are saying that the runtime will be blocked on the result for CONTAINS but not for other reduction operations, did I understand that right?

Expression `v in some_ndarray` itself isn't a blocking operation, but passing it to an if statement is. Basically, anytime you try to get an immediate Python scalar out of an expression evaluating a legate operation is a blocking operation.

> I ask this because, if we were to make an alternative implementation to figure out if we have all NaNs, that wouldn't be quite straightforward - naively, we could count the elements with NaNs along an axis and do a reduction on the count, and compare the output with the total number of elements along that axis.

Why is it not straightforward? you just need to evaluate a predicate `lambda x: is_nan(x)` on the input and perform reductions using `and`. You don't have to count the number of NaNs. 

>  now going back to the question on blocking, since we do this reduction in a non-blocking way, this is better. Is that right?

You can store the result of all-nans checks in a legate store and pass it to a downstream task that rectifies the output. all can be done in a non-blocking manner.

> but I think the backwardness that you are referring is because we know we update the output to identity, so it's really relying on that info, and that's why it fails when the array contains only identities (which fails even for some non-nan APIs, just fyi). As far as the inaccuracy is concerned, I think that's the only case where I can imagine this failing.

You should have said the only *cases*, as with axis-wise reductions, as soon as any of the rows or columns has all identities, you'd hit that case. Also, what you said is technically correct, but I don't agree with your logic. Like I said in the beginning, the whole point of this inclusion is to make cunumeric behave the same as numpy, which it fails to deliver. Let's not replace one corner case with another.

> The results are correct even without the sanitization in some of the non-nan APIs because I think they get sanitized as a consequence of what we do in fold/apply because any comparison check with NaN would fail. If we don't handle the exceptions for say, nanmin and nanmax, we might as well just call cunumeric's min and max for those two APIs (and similarly for argmin and argmax).

I don't think that's a bad thing at all. The additional guarantee of NaN variants as opposed to their normal counterparts is that NaNs are explicitly sanitized. The normal variants producing the same results as their NaN variants on some operators is just an implementation artifact, which is subject to change. The only reason that I believe NumPy's min and max return NaN when the input has a NaN is due to their implementation being sequential (i.e., they just return the result in the middle of the scanning loop as soon as they see a NaN). `nansum` and `nanprod` certainly produce different results from their normal counterparts, so I wouldn't discount the value of nan variants even as they are implemented in cuNumeric.

Reply:
> Why do you think this is an improvement or acceptable? 

yeah, here's an example using argmax (first reason):

```
x = np.array([0, 0, 0], dtype=np.uint32)
print(np.argmax(x))

# This will print array(-9223372036854775808) which is incorrect; 
# This is certainly a bug from a user perspective, but it's really an 
# artifact of our implementation, so we know it's not a bug bug.
```

that example led me to exclude the case where the array is full of identities (or a slice is full of identities) from consideration for this PR.

second, we only handle exceptions for floating point types, which means the arrays would have to be filled with values of identities that are less likely to be used, say, an array entirely with values like 1.7976931348623157e+308 for us to be false-positive. I'd guess that it's more likely to encounter cases where entire array or a slice is just NaNs as opposed to someone doing a nanmin operation on an array entirely (or a slice) filled with values like 1.7976931348623157e+308. (that's why we are not replacing one corner case with another, I think)

but I agree blocking is a concern, for sure, esp after working on torchswe. I'll see if I can remove the blocking in the next day or two; otherwise I'll remove the exception handling.

>  nansum and nanprod certainly produce different 

yeah, there is no exception handling for them, and we definitely need sanitization for them. The other four APIs are the ones that need handling of exceptions.

Reply:
>  I'll see if I can remove the blocking in the next day or two; otherwise I'll remove the exception handling.

sounds good.

> that example led me to exclude the case where the array is full of identities (or a slice is full of identities) from consideration for this PR.

the issue you pointed out is independent of what this PR is adding (though I'm not saying we shouldn't fix it), so I wouldn't use it as an excuse to not handle all identity cases correctly.

> I'd guess that it's more likely to encounter cases where entire array or a slice is just NaNs as opposed to someone doing a nanmin operation on an array entirely (or a slice) filled with values like 1.7976931348623157e+308. (that's why we are not replacing one corner case with another, I think)

I wouldn't guess the likelihood of one corner case over that of the other, and I don't like the fact that we knowingly introduce this behavior. I'd rather keep the current behavior and fix it properly in a follow-on PR.

Reply:
@magnatelee for nanarg functions, we only need to raise a ValueError when the entire array contains NaNs (or a slice contains NaNs and the axis option supplied matches the slice). There is no update to the output array unlike nanmin/nanmax. Let's say we do a bunch of collectives on the input array to find out if we encounter those two cases in a non-blocking way, w'd still need the value on python side for raising exceptions, and I think that would still block, right? It might look something like this with blocking occuring on the line before raise:

```
def nanargmin():
        ...
        if axis is not None:
            all_or_slices_nan = any(all(isnan(a), axis=axis))
        else:
            all_or_slices_nan = all(isnan(a))
        if all_or_slices_nan:
            raise ValueError(msg)
```

For the nan[min/max] functions that have NaNs in the output when a slice contains only NaNs (and similarly, if the entire array contains NaNs): let's say we update the output array in a non-blocking way to NaNs in places that need an update, but to raise an exception, we'd still block (runtime.warn just calls the warn method on warnings, so there's nothing fancy going on there) 

If that's true, it sounds like raising certain exceptions based on output would invariably lead to blocking, so unless we relax the criteria for raising exceptions in those situations, we might not get around blocking.

Reply:
@magnatelee, the inclusion check is now replaced with non-blocking operations. 

For nan[min/max], the output is updated to nan for slices that have all-NaNs using putmask. I've included the fix from Irina's PR [#980] to make sure the tests pass, but once that PR is merged-in, I'll remove that fix from this PR. RuntimeWarning is not raised to avoid blocking.

For nanarg[min/max], there is no update to the output array and a ValueError is raised if input array has an all-NaN slice or if the entire array is full of NaNs. Since this is an arg-based reductions where argument can be used downstream, I think raising a ValueError here will be useful, but it'll block the runtime.

Can you take a quick look at the PR again?